typed/reconcile_schema.go

```diff
@@ -0,0 +1,297 @@
+	}
+	element.Children.Iterate(func(pe fieldpath.PathElement) {
+		if element.Members.Has(pe) {
+			return
+		}
+		handleElement(pe, false)
+	})
+	element.Members.Iterate(func(pe fieldpath.PathElement) {
+		handleElement(pe, true)
+	})
+	return nil
+}
+
+func (v *reconcileWithSchemaWalker) doList(t *schema.List) (errs ValidationErrors) {
+	// reconcile lists changed from granular to atomic
+	if !v.isAtomic && t.ElementRelationship == schema.Atomic {
+		v.toRemove = fieldpath.NewSet(v.path) // remove all root and all children fields
+		v.toAdd = fieldpath.NewSet(v.path)    // add the root of the atomic
+		return errs
+	}
+	// reconcile lists changed from atomic to granular
+	if v.isAtomic && t.ElementRelationship == schema.Associative {
+		v.toAdd, errs = buildGranularFieldSet(v.path, v.value)
+		if errs != nil {
+			return errs
+		}
+	}
+	if v.fieldSet != nil {
+		errs = v.visitListItems(t, v.fieldSet)
+	}
+	return errs
+}
+
+func (v *reconcileWithSchemaWalker) visitMapItems(t *schema.Map, element *fieldpath.Set) (errs ValidationErrors) {
+	handleElement := func(pe fieldpath.PathElement, isMember bool) {
+		var hasChildren bool
+		tr, lookupErrs := typeRefAtPath(t, pe)
+		errs = append(errs, lookupErrs...)
+		v2 := v.prepareDescent(pe, tr)
+		v2.fieldSet, hasChildren = element.Children.Get(pe)
+		v2.isAtomic = isMember && !hasChildren
+		errs = append(errs, v2.reconcile()...)
+		v.finishDescent(v2)
+	}
+	element.Children.Iterate(func(pe fieldpath.PathElement) {
+		if element.Members.Has(pe) {
+			return
+		}
+		handleElement(pe, false)
+	})
+	element.Members.Iterate(func(pe fieldpath.PathElement) {
+		handleElement(pe, true)
+	})
+
+	return nil
+}
+
+func (v *reconcileWithSchemaWalker) doMap(t *schema.Map) (errs ValidationErrors) {
+	// reconcile maps and structs changed from granular to atomic
+	if !v.isAtomic && t.ElementRelationship == schema.Atomic {
+		if v.fieldSet != nil && v.fieldSet.Size() > 0 {
+			v.toRemove = fieldpath.NewSet(v.path) // remove all root and all children fields
+			v.toAdd = fieldpath.NewSet(v.path)    // add the root of the atomic
+		}
+		return errs
+	}
+	// reconcile maps changed from atomic to granular
+	if v.isAtomic && (t.ElementRelationship == schema.Separable || t.ElementRelationship == "") {
+		v.toAdd, errs = buildGranularFieldSet(v.path, v.value)
+		if errs != nil {
+			return errs
+		}
+	}
+	if v.fieldSet != nil {
+		errs = v.visitMapItems(t, v.fieldSet)
+	}
+	return errs
+}
+
+func buildGranularFieldSet(path fieldpath.Path, value *TypedValue) (*fieldpath.Set, ValidationErrors) {
+	result := fieldpath.NewSet(path)
+	valueFieldSet, err := value.ToFieldSet()
+	if err != nil {
+		return nil, errorf("toFieldSet: %v", err)
+	}
+	if listValueFieldSet, ok := fieldSetAtPath(valueFieldSet, path); ok {
+		valueFieldSet = prefixWithPath(path, listValueFieldSet)
+		result = result.Union(valueFieldSet)
+	}
+	return result, nil
+}
+
+func prefixWithPath(prefix fieldpath.Path, set *fieldpath.Set) *fieldpath.Set {
```

Comment:
That makes me a little sad, there's probably a better way to do that :-)

Reply:
I've switched to constructing the new field path in a more direct way that doesn't require the prefixing.